### PR TITLE
Veridream

### PR DESCRIPTION
**Pull request by hovi**

### CODE REVIEW

File: src/commonMain/kotlin/spaceEngineers/controller/CharacterControllerExtensions.kt [hunk 1]
```diff
@@ -50,6 +50,22 @@ fun Character.blockingMoveForwardByDistance(
...
 }
 
+fun Character.blockingMoveBackwardsByDistance(
+    distance: Float,
+    velocity: Float = 1f,
+    maxTries: Int = 1000,
+    startPosition: Vec3,
+): SeObservation {
+    repeat(maxTries) {
+        moveForward(velocity = -velocity).let {
```
Comment (PremekPaska) [inline src/commonMain/kotlin/spaceEngineers/controller/CharacterControllerExtensions.kt]: Why not just call the forward version with negative velocity? is this function adding anything else then just the comfort of not using negative velocity?

File: src/commonMain/kotlin/spaceEngineers/model/SeBlockDefinition.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package spaceEngineers.model
+
+
+enum class CubeSize {
+    Large, Small;
+}
```
Comment (PremekPaska) [inline src/commonMain/kotlin/spaceEngineers/model/SeBlockDefinition.kt:6]: I would refer to this as "GridSize", cube sounds more like individual cube.

I think in SE code it's called grid size, and in the UI it's called "ship size" or maybe "ship grid size".

Comment (PremekPaska) [reply src/commonMain/kotlin/spaceEngineers/model/SeBlockDefinition.kt:6]: Maybe CubeSize would work too, we should double check what's the term in SE codebase.

File: src/commonMain/kotlin/spaceEngineers/model/SeBuildProgressModel.kt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package spaceEngineers.model
+
+data class SeBuildProgressModel(
+    val buildRatioUpperBound: Float,
+    val randomOrientation: Boolean,
```
Comment (PremekPaska) [inline src/commonMain/kotlin/spaceEngineers/model/SeBuildProgressModel.kt]: what is random orientation? maybe a comment would help?-) (or a better name or both)

File: src/jvmMain/kotlin/spaceEngineers/controller/JsonRpcCharacterController.kt [hunk 4]
```diff
@@ -156,6 +155,29 @@ class JsonRpcCharacterController(
         }
     }
 
+    fun blockDefinitions(): List<SeBlockDefinition> {
+        return processParameters<Unit, List<LinkedTreeMap<Any, Any>>>(
+            params = mapOf(),
+            methodName = "Observer.BlockDefinitions"
```
Comment (PremekPaska) [inline src/jvmMain/kotlin/spaceEngineers/controller/JsonRpcCharacterController.kt]: I'm not sure about this, but shouldn't there be a verb in the method name?

Such as "getBlockDefinitions"?

File: src/jvmMain/kotlin/spaceEngineers/controller/SpaceEngineersJvmExtensions.kt [hunk 5]
```diff
@@ -1,9 +1,49 @@
...
+    items.beginUsingTool()
+    runWhileConditionUntilTimeout {
+        observer.blockById(blockId).integrity > integrity
+    }
+    items.endUsingTool()
+}
+
+fun ContextControllerWrapper.torchBackToMax(blockId: String) {
+    val maxIntegrity = observer.blockById(blockId).maxIntegrity
+    items.equip(context.torchLocation!!)
```
Comment (PremekPaska) [inline src/jvmMain/kotlin/spaceEngineers/controller/SpaceEngineersJvmExtensions.kt:42]: BTW, Joe uses "torch", but "welder" would be probably more correct...

(but it's a bit outside of the scope of this PR)

**Review by PremekPaska — APPROVED** [review]
Looks good overall.

File: src/commonMain/kotlin/spaceEngineers/model/SeBlockDefinition.kt (hunk 2 above)
Comment (hovi) [reply src/commonMain/kotlin/spaceEngineers/model/SeBlockDefinition.kt:6]: It is CubeSize in SE. I am always using same naming if possible :-)

File: src/jvmMain/kotlin/spaceEngineers/controller/JsonRpcCharacterController.kt (hunk 4 above)
Comment (hovi) [reply src/jvmMain/kotlin/spaceEngineers/controller/JsonRpcCharacterController.kt]: I understand, that mostly methods are supposed to be actions and do something, but I also really hate this "get" thing. I think in this "Definitions" interface, methods are not really doing anything, just getting the data. If I could, I'd make it property, not method to make that more clear. Does "get" really have to be in all of upcoming data methods in "Definitions"?

File: src/jvmMain/kotlin/spaceEngineers/controller/SpaceEngineersJvmExtensions.kt (hunk 5 above)
Comment (hovi) [reply src/jvmMain/kotlin/spaceEngineers/controller/SpaceEngineersJvmExtensions.kt:42]: Alright, welding/welder it is and I will start using it everywhere/slowly renaming everytime I see a torch :-D

File: src/commonMain/kotlin/spaceEngineers/model/SeBuildProgressModel.kt (hunk 3 above)
Comment (hovi) [reply src/commonMain/kotlin/spaceEngineers/model/SeBuildProgressModel.kt]: Good point, it shouldn't be here. It's random field from original SE field, that we don't need and I put it only for initial testing. Naming is same as in SE.

File: src/commonMain/kotlin/spaceEngineers/controller/CharacterControllerExtensions.kt (hunk 1 above)
Comment (hovi) [reply src/commonMain/kotlin/spaceEngineers/controller/CharacterControllerExtensions.kt]: I could have swore, there was a reason for that and that there was another difference in the method, but looks like there isn't :-D well spotted

**Pull request merged**